Code/GraphMol/SubstructLibrary/SubstructLibrarySerialization.h

```diff
@@ -0,0 +1,189 @@
+//  Copyright (c) 2019, Novartis Institutes for BioMedical Research Inc.
+//  All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Novartis Institutes for BioMedical Research Inc.
+//       nor the names of its contributors may be used to endorse or promote
+//       products derived from this software without specific prior written
+//       permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// n.b. must be included at the END of SubstructLibrary.h
+#ifndef RDK_SUBSTRUCT_LIBRARY_SERIALIZATION
+#define RDK_SUBSTRUCT_LIBRARY_SERIALIZATION
+
+#ifdef RDK_USE_BOOST_SERIALIZATION
+#include <RDGeneral/BoostStartInclude.h>
+#include <boost/archive/text_oarchive.hpp>
+#include <boost/archive/text_iarchive.hpp>
+#include <boost/serialization/vector.hpp>
+#include <boost/serialization/shared_ptr.hpp>
+#include <RDGeneral/BoostEndInclude.h>
+
+
+
+BOOST_SERIALIZATION_ASSUME_ABSTRACT(RDKit::MolHolderBase)
+BOOST_SERIALIZATION_ASSUME_ABSTRACT(RDKit::FPHolderBase)
+
+namespace boost {
+namespace serialization {
+  
+template <class Archive>
+void serialize(Archive &ar, RDKit::MolHolderBase &, const unsigned int version) {
+  RDUNUSED_PARAM(version);
+  RDUNUSED_PARAM(ar);
+}
+
+template <class Archive>
+void save(Archive &ar, const RDKit::MolHolder& molholder,
+          const unsigned int version) {
+  RDUNUSED_PARAM(version);
+  ar &boost::serialization::base_object<RDKit::MolHolderBase>(molholder);
+  
+  std::vector<std::string> pickles;
```

Comment:
This uses a fair amount of memory since you end up storing all the molecules + all the pickles.
Would it make more sense to write the pickles one-by-one to the stream instead?

Reply:
Good point.  I think there is a way to make a wrapper to “fake” the vector.  I’ll investogate.